### PR TITLE
[DIR-710] Enhancing UX for Time Input in Filters

### PR DESCRIPTION
**Pull request by julia-zielke**
## Description

The Time Input should not show the default collapsible list of numbers for hours, minutes, seconds.

## Checklist

- [x] Documentation updated if required
- [x] Test coverage is appropriate
      
## Checklist Internal

- [x] Linear issue linked (e.g. [DIR-XXXX] pull request title)
- [x] Has the PR been labeled

### CODE REVIEW

File: ui/src/design/Timepicker/timepicker-utils.tsx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  return date;
+}
+
+export function setHours(date: Date, value: string) {
+  const hours = getValidHour(value);
+  date.setHours(parseInt(hours, 10));
+  return date;
+}
+
+export type TimePickerType = "minutes" | "seconds" | "hours"; // | "12hours";
```
Comment (sebxian) [inline ui/src/design/Timepicker/timepicker-utils.tsx:105]: I think we should remove the 12hours option completely if we don't use it.

Unless there is a reason to enable this as an option for our users. But personally I would keep it simple and only offer the 24 hours format. Any thoughts on this, @stefan-kracht ?

**Review by sebxian — CHANGES_REQUESTED** [review]
Very good work, looks good and works very well.

I have some suggestions. Some of these are related to code style preferences we follow in our code base. So there's nothing wrong with your code, but it improves readability and makes our lives easier in the future if we follow the same conventions everywhere.

### Usability

Tested and works very well. But I noticed one inconvenience: When you set the time and then change the date, the time is reset to 00:00:00. It would be nice if we could preserve the time in these situations.

### Layout

image

I think it would look better with the arrow exactly aligned with the inputs.

### Add unit tests

It would be cool to add unit tests for

* `getTimeString` in `src/design/Timepicker/index.tsx`
* maybe also the helpers in `ui/src/design/Timepicker/timepicker-utils.tsx`?

This would just check if a given input produces the expected output for a few relevant cases.

You can look at `src/components/Pagination/__tests__/describePagination.test.ts` (or other files with `__tests__` in the path) for how these work, and run `yarn run test` to run them. Let me know if you need help with that.

### Consolidating duplicate files

As far as I can see, these files are almost identical:

`ui/src/pages/namespace/Events/History/components/Filters/RefineTime.tsx`
`ui/src/pages/namespace/Instances/components/Filters/RefineTime.tsx`

We should think about moving them to `ui/src/components/Filters`, so we won't have to update code in 2 (or more) places in the future. There are probably more files in the respective directories that can be consolidated, but we would have to investigate more closely (because there may also be files that are unique or specific to the Events and Instances context).

Let me know if you feel like giving this a try. I am happy to help.

File: ui/src/design/Timepicker/timepicker-utils.tsx (hunk 1 above)
Comment (julia-zielke) [reply ui/src/design/Timepicker/timepicker-utils.tsx:105]: You are right, I have to clean this file from redundant code. I would also vote for sticking with the 24 hours format. But let's see what Stefan says.

Comment (stefan-kracht) [reply ui/src/design/Timepicker/timepicker-utils.tsx:105]: I agree!

File: ui/src/design/Timepicker/index.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+const meta = {
+  title: "Components/Timepicker",
+  component: TimePicker,
+} satisfies Meta<typeof TimePicker>;
+
+export default meta;
+
+export const Default = () => {
+  const [date, setDate] = React.useState<Date>(new Date());
```
[the remark this answers is not in the document]
Comment (julia-zielke) [reply ui/src/design/Timepicker/index.stories.tsx:27]: I actually don't know anymore where I got this from, but as it does not make a difference, I changed it...

File: ui/src/design/Timepicker/timepicker-utils.tsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  return getValidNumber(value, { max: 59 });
+}
+
+type GetValidArrowNumberConfig = {
+  min: number;
+  max: number;
+  step: number;
+};
+
+export function getValidArrowNumber(
```
[the remark this answers is not in the document]
Comment (julia-zielke) [reply ui/src/design/Timepicker/timepicker-utils.tsx:67]: These functions are for the alternative input behavior via keyboard key: Arrow Up or Arrow Down
So the 'ArrowNumber' is calculated with the step size (+1 or -1).

File: ui/src/design/Timepicker/index.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+import { FC, KeyboardEventHandler } from "react";
+
+import { TimePickerInput } from "./timepicker-input";
+
+export function getTimeString(date: Date) {
+  const hours =
+    date.getHours() < 10
+      ? "0" + String(date.getHours())
+      : String(date.getHours());
```
[the remark this answers is not in the document]
Comment (julia-zielke) [reply ui/src/design/Timepicker/index.tsx:11]: That looks much better. Thanks! 🙂

**Comment by julia-zielke** [comment]
Thanks again for your suggestions! I am happy with the result now.

I think the consolidation of files and unit tests would be really valuable.
Maybe we can take a look together how this can be achieved, otherwise I would like to revisit this in a separate PR in the next release cycle, because for me it feels like a relatively big change.

Please check my code again.
We can also rename "getValidArrowNumber" and the other related functions. I just did not have an idea for a better name yet. 🙂

**Comment by sebxian** [comment]
Thanks for the updates, @julia-zielke.

I've added the unit tests and also did some refactoring that was a bit out of scope for this ticket, but it was a good opportunity do this now. I've noticed some duplicate components for filters under the events and instances routes that could be moved to our general components folder. 

@stefan-kracht, since I did some bigger refactoring it would be good to have a second peer review from you.

**Review by sebxian — APPROVED** [review]

**Pull request merged**